lib/active_merchant/billing/gateways/alelo.rb

```diff
@@ -0,0 +1,245 @@
+        })
+      end
+
+      def add_payment(post, payment, options)
+        post.merge!({
+          cardNumber: payment.number,
+          cardholderName: payment.name,
+          expirationMonth: payment.month,
+          expirationYear: format(payment.year, :two_digits),
+          securityCode: payment.verification_value
+        })
+      end
+
+      def fetch_access_token
+        params = {
+          grant_type: 'client_credentials',
+          client_id: @options[:client_id],
+          client_secret: @options[:client_secret],
+          scope: '/capture'
+        }
+
+        headers = {
+          'Accept' => 'application/json',
+          'Content-Type' => 'application/x-www-form-urlencoded'
+        }
+
+        parsed = parse(ssl_post(url('captura-oauth-provider/oauth/token'), post_data(params), headers))
+        Response.new(true, parsed[:access_token], parsed)
+      end
+
+      def remote_encryption_key(access_token)
+        response = parse(ssl_get(url('capture/key?format=json'), request_headers(access_token)))
+        Response.new(true, response[:publicKey], response)
+      end
+
+      def ensure_credentials(try_again = true)
+        multiresp = MultiResponse.new
+        access_token = @options[:access_token]
+        key = @options[:encryption_key]
+
+        if access_token.blank?
+          multiresp.process { fetch_access_token }
+          access_token = multiresp.message
+          key = nil
+        end
+
+        if key.blank?
+          multiresp.process { remote_encryption_key(access_token) }
+          key = multiresp.message
+        end
+
+        {
+          key: key,
+          access_token: access_token,
+          multiresp: multiresp.responses.present? ? multiresp : nil
+        }
+      rescue ResponseError => error
+        # retry to generate a new access_token when the provided one is expired
+        raise error unless try_again && error.response.code == '401' && @options[:access_token].present?
+
+        @options.delete(:access_token)
+        @options.delete(:encryption_key)
+        ensure_credentials false
+      end
+
+      def encrypt_payload(body, options)
+        credentials = ensure_credentials
+        key = OpenSSL::PKey::RSA.new(Base64.decode64(credentials[:key]))
+        jwk = JOSE::JWK.from_key(key)
+        alg_enc = { 'alg' => 'RSA-OAEP-256', 'enc' => 'A128CBC-HS256' }
+
+        token = JOSE::JWE.block_encrypt(jwk, body.to_json, alg_enc).compact
+
+        encrypted_body = {
+          token: token,
+          uuid: options[:uuid] || SecureRandom.uuid
+        }
+
+        return encrypted_body.to_json, credentials
+      end
+
+      def parse(body)
+        JSON.parse(body, symbolize_names: true)
+      end
+
+      def post_data(params)
+        params.map { |k, v| "#{k}=#{CGI.escape(v.to_s)}" }.join('&')
+      end
+
+      def commit(action, body, options, try_again = true)
+        payload, credentials = encrypt_payload(body, options)
```

Comment:
I might suggest extracting  the call to`ensure_credentials` to a new line above, before `encrypt_payload` for better readability. What do you think?

Reply:
yup, make sense